tests/test-cast-to-underflow.c

```diff
@@ -39,63 +39,461 @@ typedef struct
+ {__LINE__, FE_DOWNWARD, -6.475175119438025e-4966DL, -__LDBL_DENORM_MIN__},
+ {__LINE__, FE_DOWNWARD, 6.475175119438025110924438958227646e-4966DL, 0.0},
+ {__LINE__, FE_DOWNWARD, -6.475175119438025110924438958227646e-4966DL,
+  -__LDBL_DENORM_MIN__ },
+ {__LINE__, FE_DOWNWARD, 6.475175119438025110924438958227647e-4966DL,
+  __LDBL_DENORM_MIN__},
+ {__LINE__, FE_DOWNWARD, -6.475175119438025110924438958227647e-4966DL,
+  -2 * __LDBL_DENORM_MIN__},
+ {__LINE__, FE_DOWNWARD, 6.475175119438026e-4966DL, __LDBL_DENORM_MIN__},
+ {__LINE__, FE_DOWNWARD, -6.475175119438026e-4966DL, -2 * __LDBL_DENORM_MIN__},
+ {__LINE__, FE_DOWNWARD, 7e-4966DL, __LDBL_DENORM_MIN__},
+ {__LINE__, FE_DOWNWARD, -7e-4966DL, -2 * __LDBL_DENORM_MIN__},
+#endif
+ {__LINE__, FE_UPWARD, 1e-5000DL, __LDBL_DENORM_MIN__},
+ {__LINE__, FE_UPWARD, -1e-5000DL, -0.0},
+ {__LINE__, FE_UPWARD, 1e-4967DL, __LDBL_DENORM_MIN__},
+ {__LINE__, FE_UPWARD, -1e-4967DL, -0.0},
+#if __LDBL_MANT_DIG__ == 113
+ {__LINE__, FE_UPWARD, 6.475175119438025e-4966DL, __LDBL_DENORM_MIN__},
+ {__LINE__, FE_UPWARD, -6.475175119438025e-4966DL, -0.0},
+ {__LINE__, FE_UPWARD, 6.475175119438025110924438958227646e-4966DL,
+  __LDBL_DENORM_MIN__},
+ {__LINE__, FE_UPWARD, -6.475175119438025110924438958227646e-4966DL, -0.0 },
+ {__LINE__, FE_UPWARD, 6.475175119438025110924438958227647e-4966DL,
+  2 * __LDBL_DENORM_MIN__},
+ {__LINE__, FE_UPWARD, -6.475175119438025110924438958227647e-4966DL,
+  -__LDBL_DENORM_MIN__},
+ {__LINE__, FE_UPWARD, 6.475175119438026e-4966DL, 2 * __LDBL_DENORM_MIN__},
+ {__LINE__, FE_UPWARD, -6.475175119438026e-4966DL, -__LDBL_DENORM_MIN__},
+ {__LINE__, FE_UPWARD, 7e-4966DL, 2 * __LDBL_DENORM_MIN__},
+ {__LINE__, FE_UPWARD, -7e-4966DL, -__LDBL_DENORM_MIN__}
+#endif
+};
 
 /* Decimal 128 to Binary Double 64.  */
-static const d_type d128to64[] = {
+static const d128_type d128to64[] = {
+ {__LINE__, FE_TONEAREST, 1e-6176DL, 0.0},
+ {__LINE__, FE_TONEAREST, -1e-6176DL, -0.0},
  {__LINE__, FE_TONEAREST, 1e-1000DL, 0.0},
  {__LINE__, FE_TONEAREST, -1e-1000DL, -0.0},
- {__LINE__, FE_TONEAREST, 2.4703282292062327e-325DL, 0.0},
- {__LINE__, FE_TONEAREST, -2.4703282292062327e-325DL, -0.0},
- {__LINE__, FE_TONEAREST, 2.4703282292062326e-325DL, __DBL_DENORM_MIN__},
- {__LINE__, FE_TONEAREST, -2.4703282292062326e-325DL, -__DBL_DENORM_MIN__},
+ {__LINE__, FE_TONEAREST, 1e-325DL, 0.0},
+ {__LINE__, FE_TONEAREST, -1e-325DL, -0.0},
+ {__LINE__, FE_TONEAREST, 2.4703282292062326e-324DL, 0.0},
+ {__LINE__, FE_TONEAREST, -2.4703282292062326e-324DL, -0.0},
+ {__LINE__, FE_TONEAREST, 2.4703282292062327e-324DL, 0.0},
+ {__LINE__, FE_TONEAREST, -2.4703282292062327e-324DL, -0.0},
+ {__LINE__, FE_TONEAREST, 2.470328229206232720882843964341105e-324DL, 0.0},
+ {__LINE__, FE_TONEAREST, -2.470328229206232720882843964341105e-324DL, -0.0},
+ {__LINE__, FE_TONEAREST, 2.470328229206232720882843964341107e-324DL,
+  __DBL_DENORM_MIN__},
+ {__LINE__, FE_TONEAREST, -2.470328229206232720882843964341107e-324DL,
```

Comment:
@stliibm I expect this test to return 0.0. "When conversion of a value with absolute value less than or equal to half the least subnormal in the target type should produce +/- 0, not +/- the least subnormal." 

Half of the least subnormal (DBL_DENORM_MIN) = 2.4703282292062327e-324.

Since the function in trunctddf.c (see https://github.com/libdfp/libdfp/blob/master/base-math/trunctddf.c#L47) truncates the number by 17 digits the result (a_norm) of 2.470328229206232720882843964341107e-324DL is
24703282292062327 exp=323 so the result here is -0.0 since 24703282292062327 is equal half of min subnormal.

I am not sure if the correct behavior will be checking if the result is less or equal half of subnormal before truncate (maybe creating a DL with the half subnormal and compare).

Reply:
The gcc macro is defined as:
define __DBL_DENORM_MIN__ ((double)4.94065645841246544176568792868221372e-324L)
This value is not truncated to 17 digits.

If running test-cast-to-underflow on s390x which uses hardware instruction pfpo for converting between the floating-point-types,
all tests are passing (if either gcc is fixed or the rounding-mode is set with fe_dec_setround. See issue #71).

I've used the attached test to see how conversion from double to float is handling this underflow case:
half_of_flt_denorm_min=7.0064923216240854e-46
test_float_precision=9.8765430450439453
test_double_precision=9.8765432310987648634181823581457138
================================================
convert_bfp64_to_bfp32():
in = 7.0064923216240846e-46 = 0x1.fffffffffffffp-151
out= 0.0000000000000000e+00 = 0x0p+0

convert_bfp64_to_bfp32():
in = 7.0064923216240854e-46 = 0x1p-150
out= 0.0000000000000000e+00 = 0x0p+0

convert_bfp64_to_bfp32():
in = 7.0064923216240869e-46 = 0x1.0000000000001p-150
out= 1.4012984643248171e-45 = 0x1p-149

According to ieee 754-2008:
4.3 Rounding-direction attributes
Rounding takes a number regarded as infinitely precise and, if necessary, modifies it to fit in the
destination’s format while signaling the inexact exception, underflow, or overflow when appropriate (see 7).
Except where stated otherwise, every operation shall be performed as if it first produced an intermediate
result correct to infinite precision and with unbounded range, and then rounded that result according to one
of the attributes in this clause.

One attribute can be round to nearest with ties to even.
[test-underflow.txt](https://github.com/libdfp/libdfp/files/2209817/test-underflow.txt)

Reply:
> The gcc macro is defined as:
define DBL_DENORM_MIN ((double)4.94065645841246544176568792868221372e-324L)
I did not say the macro was truncated. What is being truncated is the result of:
'''
/* Get mantissa with 17 significant digits, and normalized exponent.  */
a_norm = getmantandexpd128 (a, &exp, 17, 1e17DL);
'''
In fact I was using the wrong macro before. I got the macro using 'gcc -dM -E - < /dev/null' but I've used it on my laptop (a x86_64).

Reply:
I forgot to mention: I am not blocking this PR I still think that we should approve it.